ethkey/src/extended.rs

```diff
@@ -21,6 +21,53 @@ use Public;
 use bigint::hash::{H256, FixedHash};
 pub use self::derivation::Error as DerivationError;
 
+/// Represents label that can be stored as a part of key derivation
+pub trait Label {
+	/// Length of the data that label occupies
+	fn len() -> usize;
+
+	/// Store label data to the key derivation sequence
+	/// Must not use more than `len()` bytes from slice
+	fn store(&self, target: &mut [u8]);
+}
+
+impl Label for u32 {
+	fn len() -> usize { 4 }
+
+	fn store(&self, target: &mut [u8]) {
+		use byteorder::{BigEndian, ByteOrder};
+
+		BigEndian::write_u32(&mut target[0..4], *self);
+	}
+}
+
+/// Key derivation over generic label `T`
+pub enum Derivation<T: Label> {
+	/// Soft key derivation (allow proof of parent)
+	Soft(T),
+	/// Hard key derivation (does not allow proof of parent)
+	Hard(T),
+}
+
+impl From<u32> for Derivation<u32> {
+	fn from(index: u32) -> Self {
+		if index < (2 << 30) {
+			Derivation::Soft(index)
+		}
+		else {
+			Derivation::Hard(index)
+		}
+	}
+}
+
+impl Label for H256 {
+	fn len() -> usize { 32 }
+
+	fn store(&self, target: &mut [u8]) {
+		self.copy_to(&mut target[0..4]);
```

Comment:
yeah right
i added it more like an example here, didn't payed much attention, thanx